adlfs/transfer.py

```diff
@@ -0,0 +1,312 @@
+        rate = nbytes / elapsed / 1024 / 1024
+        logger.info("Transferred %s -> %s in %f seconds at %f MB/s",
+                    src, dst, elapsed, rate)
+
+    def _update(self):
+        for (src, dst), dic in self._files.items():
+            if self._fstates[(src, dst)] == 'transferring':
+                for name in list(dic['chunks']):
+                    future = dic['chunks'][name]['future']
+                    if not future.done():
+                        continue
+                    if future.cancelled():
+                        dic['cstates'][name] = 'cancelled'
+                    elif future.exception():
+                        dic['cstates'][name] = 'errored'
+                    else:
+                        dic['cstates'][name] = 'finished'
+                if dic['cstates'].contains_all('finished'):
+                    logger.debug("Chunks transferred")
+                    chunks = list(dic['chunks'])
+                    if self._merge and len(chunks) > 1:
+                        logger.debug("Merging file: %s", self._fstates[(src, dst)])
+                        self._fstates[(src, dst)] = 'merging'
+                        dic['merge'] = self._submit(self._merge, dst, chunks)
+                    else:
+                        dic['stop'] = time.time()
+                        self._fstates[(src, dst)] = 'finished'
+                        self._status(src, dst, dic['nbytes'], dic['start'], dic['stop'])
+                elif dic['cstates'].contains_none('running'):
+                    logger.debug("Transfer failed: %s", dic['cstates'])
+                    self._fstates[(src, dst)] = 'errored'
+                else:
+                    logger.debug("Transferring chunks: %s", dic['cstates'])
+            elif self._fstates[(src, dst)] == 'merging':
+                future = dic['merge']
+                if not future.done():
+                    continue
+                if future.cancelled():
+                    self._fstates[(src, dst)] = 'cancelled'
+                elif future.exception():
+                    self._fstates[(src, dst)] = 'errored'
+                else:
+                    dic['stop'] = time.time()
+                    self._fstates[(src, dst)] = 'finished'
+                    self._status(src, dst, dic['nbytes'], dic['start'], dic['stop'])
+        self.save()
+
+    def run(self, nthreads=None, monitor=True, before_scatter=None):
+        self._nthreads = nthreads or self._nthreads
+        for src, dst in self._files:
+            self._files[(src, dst)]['start'] = time.time()
+            self._fstates[(src, dst)] = 'transferring'
+            if before_scatter:
+                before_scatter(self._adlfs, src, dst)
+            self._scatter(src, dst, self._transfer)
+        if monitor:
+            self.monitor()
+
+    def _cancel(self):
+        for dic in self._files.values():
+            for transfer in dic['chunks'].values():
+                transfer['future'].cancel()
```

Comment:
This does not kill running threads as far as I know - so network activity will continue potentially for some time.